### ISSUE TITLE
First impressions and minor things that can be fixed

### ISSUE THREAD

**Issue by AngryLoki** [syoyo/safetensors-cpp#1]
Hi, finally a native safetensors C++ parser!
And not just rust-free, also exception free and dependency free.

1. Technically, it is not header-only, it is amalgamation in header. However having a separate header and cpp file does not make build more difficult (at least for me). And having an ability to use non-amalgamated version (maybe with flag) could help in debugging loading.
2. I tried to load https://huggingface.co/runwayml/stable-diffusion-v1-5/resolve/main/v1-5-pruned-emaonly.safetensors with it and it failed with `The last tensor's data_offset.END(4264989848) must be equal to databufer size 4264989852.`. I don't know why, but this file ends with extra 4 bytes ("09 07 2F 00"). Looks like some integer in little endian. Anyway, relaxing check to `<=` fixes loading.
3. `target_link_libraries(modelreader safetensors_cpp)` propagated no-exceptions flag to modelreader, which was not desired for me. If library could consists of cpp + header, it would be very simple to stop propagation with PRIVATE flag in cmake.
4. Adding `target_include_directories(safetensors_cpp PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})` also could help to library users: it is just a common convention that `target_link_libraries` should be enough to add needed paths / flags / etc.
5. safetensors.hh should include `<cstdint>` (required for `uint8_t` in Clang on Linux)

**Comment by syoyo** [comment]
Thanks for testing!

For 2. its a good find!

https://github.com/huggingface/safetensors/blob/9610b4fcbbcf9b242e56ec679c6d1fb1d9bdb64d/safetensors/src/tensor.rs#L40
https://github.com/huggingface/safetensors/blob/9610b4fcbbcf9b242e56ec679c6d1fb1d9bdb64d/safetensors/src/tensor.rs#L315

The failure is triggered according to this requirement(The last offset must be the end of file)

The model's JSON is composed of...

```
...
"sqrt_recipm1_alphas_cumprod":{"dtype":"F32","shape":[1000],"data_offsets":[4264985848,4264989848]},"model_ema.num_updates":{"dtype":"I32","shape":[],"data_offsets":[4264989848,4264989852]}
```

So in C++, std::map sorts `model_ema.num_updates` before `sqrt_recipm1_alphas_cumprod`, so the failure occurs.

I forgot Python's dict is OrderedDict(preserves the order in which the keys are inserted, not keys are lexicographically sorted).
I may need to rewrite JSON parser to preserve its appearance order... This  may require some amount of time(and API might be changed)

**Comment by syoyo** [comment]
For 1. the header file is organized by hand to stitch over several files, so no non-amalgamated version at the moment. I may drop `header-only` description at the moment.

For 3~5, PR is much appreciated!

**Comment by syoyo** [comment]
2. is resolved via #2
